### PR TITLE
don't sync deployment when pod selector is empty

### LINKED ISSUE #23218: Don't sync daemonset or deployment when their pod selector is Everything
This is useful for most controllers but critical for daemonset and deployment due to bugs like #23170

@davidopp @bgrant0607

### PR DESCRIPTION
**Pull request by mikedanese**
Fixes #23218

cc @bgrant0607 @janetkuo

### CODE REVIEW

File: pkg/controller/deployment/deployment_controller_test.go [hunk 1]
```diff
@@ -789,3 +789,25 @@ func TestSyncDeploymentCreatesReplicaSet(t *testing.T) {
...
+	d.Spec.Selector = &empty
+	controller.dStore.Store.Add(d)
+	controller.syncDeployment(fmt.Sprintf("%s/%s", d.ObjectMeta.Namespace, d.ObjectMeta.Name))
+	if len(fake.Actions()) == 0 {
+		return
+	}
+	for _, action := range fake.Actions() {
+		t.Logf("unexpected action: %#v", action)
+	}
+	t.Errorf("expected deployment controller to not take action")
```
Comment (mikedanese) [inline pkg/controller/deployment/deployment_controller_test.go]: On master this test fails with:
```
=== RUN TestDeploymentController_dontSyncDeploymentsWithEmptyPodSelector
--- FAIL: TestDeploymentController_dontSyncDeploymentsWithEmptyPodSelector (0.00s)
        deployment_controller_test.go:810: unexpected action: core.CreateActionImpl{ActionImpl:core.ActionImpl{Namespace:"default", Verb:"create", Resource:"replicasets", Subresource:""}, Object:(*extensions.ReplicaSet)(0xc2081238c0)}
        deployment_controller_test.go:810: unexpected action: core.UpdateActionImpl{ActionImpl:core.ActionImpl{Namespace:"default", Verb:"update", Resource:"deployments", Subresource:""}, Object:(*extensions.Deployment)(0xc208081b80)}
        deployment_controller_test.go:810: unexpected action: core.UpdateActionImpl{ActionImpl:core.ActionImpl{Namespace:"", Verb:"update", Resource:"replicasets", Subresource:""}, Object:(*extensions.ReplicaSet)(0xc208123b00)}
        deployment_controller_test.go:810: unexpected action: core.ListActionImpl{ActionImpl:core.ActionImpl{Namespace:"", Verb:"list", Resource:"pods", Subresource:""}, ListRestrictions:core.ListRestrictions{Labels:labels.nothingSelector{}, Fields:fields.andTerm{}}}
        deployment_controller_test.go:810: unexpected action: core.UpdateActionImpl{ActionImpl:core.ActionImpl{Namespace:"default", Verb:"update", Resource:"deployments", Subresource:"status"}, Object:(*extensions.Deployment)(0xc2081b6500)}
        deployment_controller_test.go:812: expected deployment controller to not take action
FAIL
```

File: pkg/controller/deployment/deployment_controller.go [hunk 2]
```diff
@@ -426,6 +426,11 @@ func (dc *DeploymentController) syncDeployment(key string) error {
 	}
 
 	d := obj.(*extensions.Deployment)
+	everything := unversioned.LabelSelector{}
+	if reflect.DeepEqual(d.Spec.Selector, &everything) {
+		dc.eventRecorder.Eventf(d, api.EventTypeWarning, "SelectingAll", "This controller is selecting all pods. Skipping sync.")
```
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: s/controller/Deployment/

I doubt "skipping sync" is very meaningful to users. How about "A non-empty selector is required."

**Comment by mikedanese** [comment]
@bgrant0607 @janetkuo I fixed both the daemon set and the deployment event message. PTAL

File: pkg/controller/deployment/deployment_controller_test.go (hunk 1 above)
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller_test.go]: Clarification: WIthout the controller change, this test fails, as expected.

File: pkg/controller/deployment/deployment_controller.go [hunk 3]
```diff
@@ -426,6 +426,11 @@ func (dc *DeploymentController) syncDeployment(key string) error {
 	}
 
 	d := obj.(*extensions.Deployment)
+	everything := unversioned.LabelSelector{}
+	if reflect.DeepEqual(d.Spec.Selector, &everything) {
+		dc.eventRecorder.Eventf(d, api.EventTypeWarning, "SelectingAll", "This deployment is selecting all pods. A non-empty sync is required.")
```
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: s/sync/selector/

**Comment by bgrant0607** [comment]
Thanks much. LGTM

**Pull request merged**